### ISSUE TITLE
nyc reports no coverage in parallel mode in powershell/cmd

### ISSUE THREAD

**Issue by curbengh** [mochajs/mocha#4372]
Cross-posting https://github.com/istanbuljs/nyc/issues/1328 here.

### Prerequisites

- [x] Checked that your issue hasn't already been filed by cross-referencing [issues with the `faq` label](https://github.com/mochajs/mocha/issues?utf8=%E2%9C%93&q=is%3Aissue%20label%3Afaq%20)
- [ ] Checked next-gen ES issues and syntax problems by using the same environment and/or transpiler configuration without Mocha to ensure it isn't just a feature that actually isn't supported in the environment in question or a bug in your code.
- [x] 'Smoke tested' the code to be tested by running it outside the real test suite to get a better sense of whether the problem is in the code under test, your usage of Mocha, or Mocha itself
- [x] Ensured that there is no discrepancy between the locally and globally installed versions of Mocha. You can find them with: `node node_modules/.bin/mocha --version`(Local) and `mocha --version`(Global). We recommend that you _not_ install Mocha globally.

## Link to bug demonstration repository
https://github.com/hexojs/hexo/pull/4374

## Expected Behavior
`nyc --reporter=lcovonly mocha --parallel` should produce "./coverage/lcov.info".
`nyc --reporter=text-summary mocha --parallel` should display coverage report.

## Observed Behavior

`nyc --reporter=lcovonly mocha --parallel` generates an **empty** "./coverage/lcov.info".
`nyc --reporter=text-summary mocha --parallel`:

```
=============================== Coverage summary ===============================
Statements   : Unknown% ( 0/0 )
Branches     : Unknown% ( 0/0 )
Functions    : Unknown% ( 0/0 )
Lines        : Unknown% ( 0/0 )
================================================================================
```

### Troubleshooting steps
No issue in mocha's default serial mode.

## Environment Information

```
  System:
    OS: Linux 4.9 Solus 4.1
    CPU: (1) x64 Intel(R) Core(TM) i7-4500U CPU @ 1.80GHz
    Memory: 1.53 GB / 2.94 GB
  npmPackages:
    nyc: ^15.0.0 => 15.1.0 

```

**Reproduces how often:** [What percentage of the time does it reproduce?] 100%

### Versions

- The output of `mocha --version` and `node node_modules/.bin/mocha --version`: 8.0.1 & 8.0.1
- The output of `node --version`: v12.18.0
- Your shell (e.g., bash, zsh, PowerShell, cmd): bash
- Your browser and version (if running browser tests): NA
- Any third-party Mocha-related modules (and their versions): NA
- Any code transpiler (e.g., TypeScript, CoffeeScript, Babel) being used (and its version): NA

### Additional Information

".mocharc.yml"
``` yml
color: true
reporter: spec
ui: bdd
full-trace: true
exit: true
require: 
  - "chai/register-should"
parallel: true
```

**Comment by sgissinger** [comment]
Hi,
I have the same behavior, everything is good when running `nyc mocha`.

But when running `nyc mocha --parallel` the result is the same as what described @curbengh

**Comment by Bluefinger** [comment]
I can confirm I have this issue as well. My setup is with pnpm + typescript, and while I have to run nyc with `--use-spawn-wrap`, the moment I run mocha in parallel mode, I lose my code coverage. Running mocha with parallel off gets me working coverage from nyc.

Env: Windows 10
Mocha version: 8.1.1
Node version: v14.4.0
Shell: PowerShell
Transpiler: Typescript @ v3.9.7

"`.mocharc.js`":
```js
module.exports = {
  checkLeaks: true,
  diff: true,
  color: true,
  extension: ["ts"],
  growl: false,
  reporter: "spec",
  recursive: false,
  require: ["ts-node/register"],
  spec: "test/**/*.spec.ts",
  slow: 500,
  // parallel: !process.env.CI,
};
```

**Comment by mdrichardson** [comment]
I'm running into this as well. Has anybody found a temporary workaround?

**Comment by curbengh** [comment]
I disable parallel mode in coverage test.

```
  "scripts": {
    "test": "mocha test/index.js",
    "test-cov": "nyc --reporter=lcovonly npm run test -- --no-parallel",
  },
```

**Comment by outsideris** [comment]
I'm not sure whether it is mocha's bug. I tested it  with hexo@`ebc9008f` 

Results are different base on hexo's test suites. I did dig hexo source code, but I didn't find what affect coverage.

### test/scripts/console/index.js
```
$ ./node_modules/.bin/nyc --reporter=text-summary ./node_modules/.bin/mocha --parallel test/scripts/console/index.js

=============================== Coverage summary ===============================
Statements   : Unknown% ( 0/0 )
Branches     : Unknown% ( 0/0 )
Functions    : Unknown% ( 0/0 )
Lines        : Unknown% ( 0/0 )
================================================================================
```

### test/scripts/extend/index.js
```
$ ./node_modules/.bin/nyc --reporter=text-summary ./node_modules/.bin/mocha --parallel test/scripts/extend/index.js

=============================== Coverage summary ===============================
Statements   : 97.2% ( 347/357 )
Branches     : 92.72% ( 191/206 )
Functions    : 100% ( 81/81 )
Lines        : 99.03% ( 306/309 )
================================================================================
```

**Comment by boneskull** [comment]
need to find the root cause here... I'll take a look at this today.  I'll be using `botbuilder-js` for an example repo (@mdrichardson)

**Comment by boneskull** [comment]
@mdrichardson Running `nyc` with my working copy of Mocha's `master` branch results in coverage, but running it with `mocha@v8.1.3` (the current version) does not.  trying to figure out why

**Comment by mdrichardson** [comment]
@boneskull Ooooh, that's interesting. I tried it again a couple of weeks ago with the latest mocha/nyc at the time and it didn't work.

**Comment by boneskull** [comment]
tbh I'm a little confused about the monorepo setup. how to upgrade mocha on all the subpackages? I also note that nyc is present in the subpackages' package.json but not the monorepo package.json

**Comment by mdrichardson** [comment]
@boneskull I just did a find/replace w/ regex, as I recall.

**Comment by boneskull** [comment]
wow.

this was fixed by the upgrade to `workerpool` in #4465 which included my PR josdejong/workerpool#176.  I imagine `nyc` was getting confused due to temporarily-orphaned processes, somehow.

(@mdrichardson mind you, in `botbuilder-js`, there is a bit of weirdness that you'll need to look into when running in parallel)

so `nyc` and parallel mode will work properly in `mocha` v8.2.0, which I'm getting ready for release.  early next week at latest, today soonest

**Comment by boneskull** [comment]
closed by #4465

**Comment by mdrichardson** [comment]
@boneskull Awesome. Nice find!

...and yeah, I'm aware of the weirdness ;)

**Comment by boneskull** [comment]
@mdrichardson you may want to `--trace-warnings` and open an issue here if mocha is the culprit.  I've seen mocha use too many event listeners on occasion when running in parallel, and I'd like to figure out what's going on there if it's still a problem.

**Comment by mdrichardson** [comment]
@boneskull Will do. I believe the event listeners piece, at least [is unrelated](https://github.com/microsoft/botbuilder-js/pull/2699)

**Comment by mdrichardson** [comment]
@boneskull I still don't seem to get coverage reporting with [`botbuilder-js`](https://github.com/microsoft/botbuilder-js), when enabling `--parallel`. Here are the steps to reproduce:

1. `git clone https://github.com/microsoft/botbuilder-js`
2. `cd botbuilder-js`
3. Open in VS Code and Global Find/Replace (regex):
    1. `"mocha":\s?"\^?\d\d?\.\d\.?\d?",` with `"mocha": "^8.2.0",`
    2. `"nyc":\s?"\^?\d\d?\.\d\.?\d?"`, with `"nyc": "^15.1.0",`
4. Just in case, `npm i -G nyc@latest mocha@latest`
5. `npm i`
6. `npm run build`
7. Add `--parallel` to `package.json > scripts > test-coverage`
8. `npm run test-coverage`

Am I missing something?

**Comment by boneskull** [comment]
@mdrichardson I'll take a look

**Comment by boneskull** [comment]
1. clone botbuilder
1. search/replace as you did for mocha (I didn't bother upgrading nyc)
1. I ran `lerna clean -y` then `rm -rf node_modules`
1. now `npm i` and `npm run build`
1. update `test-coverage` to `nyc mocha \"libraries/@(adaptive*|bot*)/tests/**/*.test.js\" --exit --check-leaks --parallel

**Comment by mdrichardson** [comment]
@boneskull Alright, I'll give that a go. Thanks again!!!!

**Comment by boneskull** [comment]
I can't make it break... maybe there's something else about your setup. what node/npm version, os?

**Comment by mdrichardson** [comment]
Yeah, I suspected it could be something like that. Testing now, but:

- Windows
- Node v12.16.1 (I think worker threads are still in beta here?)
- npm v 6.13.4

**Comment by mdrichardson** [comment]
@boneskull Hmm...I tried everything again w/ Node v12.16.1 (original), v12.19.0, and v15.0.0 (and relevant npm version) and still don't get the coverage report. I also went in and deleted `%appdata%/npm/node_modules`.

I know if you can't repro, it's pretty much impossible for you to do much about it. Any troubleshooting steps you'd recommend on my end?

**Comment by boneskull** [comment]
I'll try it on Windows.  using powershell or something else?

**Comment by mdrichardson** [comment]
I was using command prompt, but I'll try bash and powershell

**Comment by mdrichardson** [comment]
@boneskull Worked in bash, but not command prompt or Powershell

**Comment by boneskull** [comment]
OK, that's good to know. I can reproduce in Powershell.  is the problem mocha's or nyc's?  hmmm...

**Comment by boneskull** [comment]
I've narrowed it down; see https://github.com/boneskull/mocha-nyc-parallel-issue and the README there

`mocha` is not actually used in that repo; it's not needed to reproduce.  In short: `child_process.fork()` causes no coverage to be collected (`mocha` forks worker processes by way of `workerpool`).  I'm not sure if this is something `nyc` can fix, but I'll dig around.  `workerpool` needs IPC, so I don't see it being able to avoid the fork.

**Comment by boneskull** [comment]
OK, I've updated the `README` in that repo and added an example of `fork()` which will work to collect coverage.  This would be an upstream fix.

**Comment by boneskull** [comment]
I'll send a PR to `workerpool` to try to address the issue.

**Comment by boneskull** [comment]
also see istanbuljs/nyc#762

**Comment by tal-sapan** [comment]
Hi @boneskull, since the fix in `workerpool` was merged and a version was released, could you please update the dependency in `mocha` to fix this issue?

Thanks!

**Comment by juergba** [comment]
The fix in `workerpool` was released with v6.0.3, Mocha uses v6.1.0.
Closing this issue.
